### PR TITLE
Use linecache.lazycache to cache generated code.

### PR DESCRIPTION
**Pull request by davides**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63453

Summary:
Instead of patching linecache.getlines, use linecache.lazycache and
parts of the loader protocol described in PEP-302

Test Plan:
python3 test/test_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30388176](https://our.internmc.facebook.com/intern/diff/D30388176)

### CODE REVIEW

**Comment by davides** [comment]
@davides has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388176).

File: torch/fx/graph_module.py [hunk 1]
```diff
@@ -15,28 +15,60 @@
...
+        linecache.lazycache(key, globals_copy)
+
+        return key
+
+    # Part of the loader protocol (PEP 302)
+    # linecache will use this method when trying to find source code
+    def get_source(self, module_name) -> str:
+        if module_name in self.eval_cache:
+            return self.eval_cache[module_name]
+        return None
```
Comment (suo) [inline torch/fx/graph_module.py:58]: this is causing mypy to fail—looks like PEP302 implies it's an optional[str]

**Review by suo — CHANGES_REQUESTED** [review]
Needs a small fix but overall looks pretty good!

File: torch/fx/graph_module.py (hunk 1 above)
Comment (davides) [reply torch/fx/graph_module.py:58]: Good catch, updated.

**Review by suo — APPROVED** [review]
lgtm, provided tests pass!

**Comment by suo** [comment]
oops, I accepted a little too early, see comment on https://github.com/pytorch/pytorch/pull/63497

**Comment by davides** [comment]
@davides has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388176).

**Comment by davides** [comment]
@davides has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388176).

**Pull request closed without merging**